include/swift/AST/DiagnosticsSema.def

```diff
@@ -3011,7 +3011,11 @@ WARNING(optional_pattern_match_promotion,none,
         "pattern match introduces an implicit promotion from %0 to %1",
         (Type, Type))
 WARNING(optional_to_any_coercion,none,
-        "expression implicitly coerced from %0 to %1", (Type, Type))
+        "coercion of implicitly unwrappable value of type %0 to %1 does not "
```

Comment:
This needs to be a separate diagnostic to `optional_to_any_coercion` that we diagnose specifically for IUOs, as we still want to emit `expression implicitly coerced from 'T?' to 'Any'` for plain non-IUO optional-to-Any coercions.

Reply:
Fixed.

Reply:
@theblixguy Thanks! When you get round to updating the tests, could you please add the following test case (from [my previous comment](https://github.com/apple/swift/pull/23617#issuecomment-481885817)):

```swift
let fn: (() -> Int?)! = { nil }
takeAny(fn())
```

as I believe that will incorrectly trigger the IUO diagnostic rather than the plain optional one. It's up to you whether you want to fix the logic in `hasImplicitlyUnwrappedResult` in this PR – if not, you can just leave a FIXME on it.

Reply:
Yes indeed it triggers the IUO diagnostic. Maybe we can check if the return type of a function or subscript is optional or not, besides the IUO attribute on the decl?

Reply:
@theblixguy The problem with checking the return type is that it'll be optional for both IUOs and regular optionals. How about checking that when you look through an `ApplyExpr`, that the decl you end up with is an `AbstractFunctionDecl`? e.g:

```cxx
    static bool hasImplicitlyUnwrappedResult(Expr *E) {
      auto *decl = getDeclForImplicitlyUnwrappedAttr(E);
      return decl
        && decl->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>();
    }

    static ValueDecl *getDeclForImplicitlyUnwrappedAttr(Expr *E) {
      E = E->getValueProvidingExpr();
      
      // Look through implicit conversions like loads, derived-to-base
      // conversion, etc.
      if (auto *ICE = dyn_cast<ImplicitConversionExpr>(E))
        E = ICE->getSubExpr();

      if (auto *subscript = dyn_cast<SubscriptExpr>(E)) {
        if (subscript->hasDecl())
          return subscript->getDecl().getDecl();

        return nullptr;
      }

      if (auto *memberRef = dyn_cast<MemberRefExpr>(E))
        return memberRef->getMember().getDecl();

      if (auto *declRef = dyn_cast<DeclRefExpr>(E))
        return declRef->getDecl();

      // Look through an apply to see if we're calling a function decl, which
      // might have an IUO attribute indicating that the return value is an IUO.
      if (auto *apply = dyn_cast<ApplyExpr>(E)) {
        auto *decl = apply->getCalledValue();
        if (decl && isa<AbstractFunctionDecl>(decl))
          return decl;
      }
      return nullptr;
    }
```